src/Workspaces/Core/Portable/Utilities/ValuesSources/ValueSource.cs

```diff
@@ -1,27 +1,36 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+#nullable enable
+
+using System.Diagnostics.CodeAnalysis;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
 
 namespace Roslyn.Utilities
 {
     /// <summary>
-    /// A class that abstracts the accessing of a value 
+    /// A class that abstracts the accessing of a value that is guaranteed to be available at some point.
     /// </summary>
     internal abstract class ValueSource<T>
     {
-        public abstract bool TryGetValue(out T value);
+        public abstract bool TryGetValue([MaybeNullWhen(false)]out T value);
         public abstract T GetValue(CancellationToken cancellationToken = default);
         public abstract Task<T> GetValueAsync(CancellationToken cancellationToken = default);
+    }
+
+    internal static class ValueSourceExtensions
+    {
+        internal static T? GetValueOrNull<T>(this Optional<T> optional) where T : class
+            => optional.Value;
+
+        internal static T GetValueOrDefault<T>(this Optional<T> optional) where T : struct
+            => optional.Value;
```

Comment:
💡 Consider moving these to a new extension methods class `OptionalExtensions`

Reply:
I'll follow up and get these moved to the compiler.